pkg/agent/client/client.go

```diff
@@ -0,0 +1,213 @@
+package client
+
+import (
+	"bytes"
+	"context"
+	"crypto/ecdsa"
+	"crypto/tls"
+	"crypto/x509"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"net/url"
+	"sync"
+	"time"
+
+	"github.com/sirupsen/logrus"
+	spiffe_tls "github.com/spiffe/go-spiffe/tls"
+	"github.com/spiffe/spire/pkg/common/grpcutil"
+	"github.com/spiffe/spire/pkg/common/util"
+	"github.com/spiffe/spire/proto/api/node"
+	"github.com/spiffe/spire/proto/common"
+
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+)
+
+var (
+	ErrPartialResponse   = errors.New("partial response received")
+	ErrUnableToGetStream = errors.New("unable to get a stream")
+)
+
+type Client interface {
+	FetchUpdates(req *node.FetchSVIDRequest) (*Update, error)
+
+	// Release releases any resources that were held by this Client, if any.
+	Release()
+}
+
+// Config holds a client configuration
+type Config struct {
+	Addr        net.Addr
+	Log         logrus.FieldLogger
+	TrustDomain url.URL
+	// KeysAndBundle is a callback that must return the keys and bundle used by the client
+	// to connect via mTLS to Addr.
+	KeysAndBundle func() (*x509.Certificate, *ecdsa.PrivateKey, []*x509.Certificate)
+}
+
+type Update struct {
+	Entries map[string]*common.RegistrationEntry
+	SVIDs   map[string]*node.Svid
+	Bundle  []byte
+}
+
+type client struct {
+	c    *Config
+	conn *grpc.ClientConn
+	m    sync.Mutex
+	// Callback to be used for testing purposes.
+	newNodeClientCallback func() (node.NodeClient, error)
+}
+
+// New creates a new client struct with the configuration provided
+func New(c *Config) *client {
+	return &client{
+		c: c,
+	}
+}
+
+func newGRPCConn(c *Config) (*grpc.ClientConn, error) {
+	credFunc := func() (credentials.TransportCredentials, error) {
```

Comment:
Perhaps this can be defined as a dedicated function? Would definitely make `newGRPCConn` easier to read